pep-0634.rst

```diff
@@ -0,0 +1,706 @@
+
+  match get_shape():
+      case Line(start := Point(x, y), end) if start == end:
+          print(f"Zero length line at {x}, {y}")
+
+The name on the left of the walrus operator can be used in a guard, in
+the case block, or after the ``match`` statement.  However, the name will
+*only* be bound if the sub-pattern succeeds.  Another example::
+
+  match group_shapes():
+      case [], [point := Point(x, y), *other]:
+          print(f"Got {point} in the second group")
+          process_coordinates(x, y)
+          ...
+
+Technically, most such examples can be rewritten using guards and/or nested
+``match`` statements, but this will be less readable and/or will produce less
+efficient code. Essentially, most of the arguments in PEP 572 apply here
+equally.
+
+The wildcard ``_`` is not a valid name here.
+
+
+.. _runtime:
+
+Runtime specification
+=====================
+
+The Match Protocol
+------------------
+
+The equivalent of an ``isinstance`` call is used to decide whether an
+object matches a given class pattern and to extract the corresponding
+attributes.  Classes requiring different matching semantics (such as
+duck-typing) can do so by defining ``__instancecheck__`` (a
+pre-existing metaclass hook) or by using ``typing.Protocol``.
+
+The procedure is as following:
+
+* The class object for ``Class`` in ``Class(<sub-patterns>)`` is
+  looked up and ``isinstance(obj, Class)`` is called, where ``obj`` is
+  the value being matched.  If false, the match fails.
+
+* Otherwise, if any sub-patterns are given in the form of positional
+  or keyword arguments, these are matched from left to right, as
+  follows.  The match fails as soon as a sub-pattern fails; if all
+  sub-patterns succeed, the overall class pattern match succeeds.
+
+* If there are match-by-position items and the class has a
+  ``__match_args__`` attribute, the item at position ``i``
+  is matched against the value looked up by attribute
+  ``__match_args__[i]``. For example, a pattern ``Point2d(5, 8)``,
+  where ``Point2d.__match_args__ == ["x", "y"]``, is translated
+  (approximately) into ``obj.x == 5 and obj.y == 8``.
+
+* If there are more positional items than the length of
+  ``__match_args__``, a ``TypeError`` is raised.
+
+* If the ``__match_args__`` attribute is absent on the matched class,
+  and one or more positional item appears in a match,
+  ``TypeError`` is also raised. We don't fall back on
+  using ``__slots__`` or ``__annotations__`` -- "In the face of ambiguity,
+  refuse the temptation to guess."
+
+* If there are any match-by-keyword items the keywords are looked up
+  as attributes on the subject.  If the lookup succeeds the value is
+  matched against the corresponding sub-pattern.  If the lookup fails,
+  the match fails.
+
+Such a protocol favors simplicity of implementation over flexibility and
+performance. For other considered alternatives, see "extended matching".
+
+For the most commonly-matched built-in types (``bool``,
```

Comment:
Should we remove `bool` from this list? `True` and `False` now match by identity and `bool` can't be subclassed, so its just confusing here.

(It will still work, though, since it inherits this behavior from `int`.)

Reply:
Actually, I'll leave it and add `bool(b)` as one of the examples below. It's a valid use-case that belongs in a language spec, I think.